### PR TITLE
fix mkl batch gemm

### PR DESCRIPTION
**Pull request by TaoLv**
Since the same m/n/k is used for all single gemms, so we can put all these gemms into one group of mkl batch gemm.

@yajiedesign @piiswrong please review again.

### CODE REVIEW

**Comment by TaoLv** [comment]
@sxjscience

**Comment by sxjscience** [comment]
Does it further improves the performance?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by TaoLv** [comment]
@sxjscience I think performance should be same and I have verified that from mxnet level. The main purpose of this PR is for code refine and reducing some memory usage.

**Comment by sxjscience** [comment]
Yeah. I think it’s a nice improvement.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: mshadow/dot_engine-inl.h [hunk 1]
```diff
@@ -293,40 +293,45 @@ struct BLASEngine<cpu, float> {
...
   }
 
-    cblas_sgemm_batch(CblasColMajor, p_transa.data(), p_transb.data(),
-                      p_m.data(), p_n.data(), p_k.data(),
-                      p_alpha.data(), pp_A.data(), p_lda.data(), pp_B.data(),
-                      p_ldb.data(), p_beta.data(), pp_C.data(), p_ldc.data(),
-                      1, p_group_sizeb.data());
+  cblas_sgemm_batch(CblasColMajor, p_transa, p_transb,
+                    p_m, p_n, p_k, p_alpha, pp_A.data(), p_lda, pp_B.data(),
+                    p_ldb, p_beta, pp_C.data(), p_ldc, 1, p_group_sizeb);
```
Comment (sxjscience) [inline mshadow/dot_engine-inl.h]: Why not simply using &m, &n, &k for p_m, p_n, p_k?

**Comment by xinyu-intel** [comment]
This commit may cause perf regression and we are working on it. Please do not merge now. Thanks！

File: mshadow/dot_engine-inl.h [hunk 2]
```diff
@@ -400,40 +405,45 @@ struct BLASEngine<cpu, double> {
...
-  std::vector<int> p_lda(batch_count, lda);
-  std::vector<int> p_ldb(batch_count, ldb);
-  std::vector<int> p_ldc(batch_count, ldc);
-  std::vector<double> p_alpha(batch_count, alpha);
-  std::vector<double> p_beta(batch_count, beta);
-  std::vector<const double*> pp_A;
-  std::vector<const double*> pp_B;
-  std::vector<double*> pp_C;
+  // since same m/n/k is used for all single gemms, so we put all gemms into one group
+  // group size = 1
```
Comment (rongzha1) [inline mshadow/dot_engine-inl.h]: group count = 1,  group size = batch_count,   so p_m, p_n   need to use MKL_INT p_m[batch_count]  that is std::vector<int> p_k(batch_count, k);

File: mshadow/dot_engine-inl.h [hunk 3]
```diff
@@ -293,40 +293,45 @@ struct BLASEngine<cpu, float> {
...
-  std::vector<CBLAS_TRANSPOSE> p_transa(batch_count, cblas_a_trans);
-  std::vector<CBLAS_TRANSPOSE> p_transb(batch_count, cblas_b_trans);
+  MKL_INT p_group_sizeb[1] = {batch_count};
+  CBLAS_TRANSPOSE p_transa[1] = {cblas_a_trans};
+  CBLAS_TRANSPOSE p_transb[1] = {cblas_b_trans};
+
+  std::vector<const float*> pp_A;
+  std::vector<const float*> pp_B;
+  std::vector<float*> pp_C;
+  pp_A.reserve(batch_count);
```
Comment (rongzha1) [inline mshadow/dot_engine-inl.h]: no need reserve for pp_A pp_B pp_C

**Comment by TaoLv** [comment]
Talked with @xinyu-intel offline. His performance regression was caused by other changes and not related to this PR. Code change here has passed MXNet operator unit test: https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_operator.py#L2576.

I also used the following code to verify the performance change and got similar results before and after this PR:
```
import mxnet as mx
import numpy as np
import time

x = mx.sym.Variable('x')
y = mx.sym.Variable('y')
input_x = np.random.rand(200, 256, 256)
input_y = np.random.rand(200, 256, 512)

sym = mx.symbol.batch_dot(x, y).bind(mx.cpu(), {'x': mx.nd.array(input_x), 'y': mx.nd.array(input_y)})

start = time.time()
for i in range(1010):
    if i == 10:
        start = time.time()
    sym.forward(is_train=False)[0].asnumpy()

print((time.time() - start))
```

Is it okay to merge? @sxjscience @piiswrong

**Review by sxjscience — APPROVED** [review]

File: mshadow/dot_engine-inl.h (hunk 1 above)
Comment (TaoLv) [reply mshadow/dot_engine-inl.h]: To make the code easier to understand. `MKL_INT p_m[1] = {m};` means this batched GEMM only has one group and all GEMMs in this group share the same `m` value. Maybe in the future, we can extend this to `MKL_INT p_m[2] = {m1, m2};`. Then we have two groups in one batched GEMM and the first group has `m1` while the second group has `m2`. Using `&m` in this API will hide this definition and make it a little confusing.

**Comment by TaoLv** [comment]
@sxjscience @piiswrong @eric-haibin-lin is this good to merge?

**Comment by pengzhao-intel** [comment]
@TaoLv could you rebase the code and run CI again?

**Comment by eric-haibin-lin** [comment]
Is there reference PR in MXNet to test the mshadow change end2end?

**Pull request merged**